File: packages/sites/src/resource-catalog.ts
import {
  canonical,
  hash,
  createBundle,
  verifyBundle,
  verifyStoredBundle,
  decryptStoredBundle,
  type Bundle,
  type Identity,
  type PublicIdentity,
} from "../../core/src/index";
import { validateIdentity } from "../../core/src/index";
import {
  RegistryIntegrityError,
  type RegistryTransaction,
} from "../../groups/src/storage";
import {
  describeSiteResource,
  parseSiteResource,
  type SiteReadScope,
} from "../../content/src/site-resource";
import {
  createResourceOperations,
  type ResourceCreationRecord,
  type ResourceOperation,
} from "./resource-operations";
import { SitePrivateRecords } from "./private-storage";

interface Database {
  transaction<T>(callback: (tx: RegistryTransaction) => T): T;
}
export interface ResourceOperationHandle {
  sequence: number;
  operationId: string;
  fingerprint: string;
}
const registry = createResourceOperations({ hash });
const same = (a: unknown, b: unknown) => canonical(a) === canonical(b);
function insist(ok: unknown, reason: string): asserts ok {
  if (!ok) throw new RegistryIntegrityError(reason);
}

/** Signing-owned local creation journal. Transactions return only after real
 * commit. The runtime copies authorized bytes to ContentStore without broadcast,
 * verifies the copy and calls ready. Lost responses recover the retained ID. */
export class NodeResourceCatalog {
  private readonly key: string;
  constructor(
    private readonly database: Database,
    private readonly identity: Identity,
  ) {
    this.key =
      "resource:" +
      hash(
        canonical({
          domain: "relayloom/resource-catalog-key/1",
          owner: identity.public.id,
        }),
      );
  }
  private verify(bundle: Bundle, op: ResourceOperation) {
    verifyStoredBundle(bundle);
    insist(
      bundle.manifest.kind === "site-resource" &&
        bundle.manifest.author.id === this.identity.public.id,
      "Autor ou tipo do recurso preparado inválido",
    );
    const resource = parseSiteResource(
      decryptStoredBundle(bundle, this.identity),
    );
    const scope: SiteReadScope =
      bundle.manifest.publicKey !== null
        ? "public"
        : bundle.manifest.keys.map((k) => k.reader).sort();
    const reference = describeSiteResource(resource, {
      id: bundle.manifest.id,
      authorId: bundle.manifest.author.id,
      kind: bundle.manifest.kind,
    });
    insist(
      same(reference, op.reference) &&
        same(scope, op.recipients) &&
        bundle.manifest.created === op.created &&
        bundle.manifest.expires === op.expires &&
        hash(canonical(bundle)) === op.bundleHash,
      "Preparação não corresponde ao registo de criação",
    );
    return JSON.parse(canonical(bundle)) as Bundle;
  }
  private run<T>(
    fn: (
      values: SitePrivateRecords,
      record: ResourceCreationRecord,
      stage: Bundle | null,
    ) => T,
  ): T {
    return this.database.transaction((tx) =>
      SitePrivateRecords.runResource(tx, this.identity, (values) => {
        const saved = values.read(this.key + ":record"),
          raw = values.read(this.key + ":stage");
        let record: ResourceCreationRecord;
        if (saved === null) {
          insist(
            tx.keys("resource:").length === 0,
            "Registo de criação ausente com dados existentes",
          );
          record = registry.initial(this.identity.public.id);
        } else {
          try {
            record = registry.validate(saved, this.identity.public.id);
          } catch {
            throw new RegistryIntegrityError(
              "Registo de criação de recursos inválido",
            );
          }
        }
        const pending = record.operations.find(
          (op) => op.phase === "copy-pending",
        );
        let stage: Bundle | null = null;
        if (pending) {
          insist(raw !== null, "Preparação de recurso em falta");
          try {
            stage = this.verify(raw as Bundle, pending);
          } catch {
            throw new RegistryIntegrityError("Preparação de recurso inválida");
          }
          if (pending.expires <= Date.now()) {
            record = registry.expire(
              record,
              this.identity.public.id,
              Date.now(),
            );
            values.remove(this.key + ":stage");
            values.write(this.key + ":record", record);
            stage = null;
          }
        } else
          insist(raw === null, "Assinatura de recurso sem operação pendente");
        return fn(values, record, stage);
      }),
    );
  }
  state() {
    return this.run((_values, record) => structuredClone(record));
  }
  operation(sequence: number, operationId: string) {
    return this.run((_values, record) => {
      const found = registry.lookup(
        record,
        this.identity.public.id,
        sequence,
        operationId,
      );
      return { operation: found.operation, retired: found.retired };
    });
  }
  prepare(
    input: unknown,
    resolveReaders: (scope: SiteReadScope) => PublicIdentity[] | "public",
  ) {
    const checked = registry.request(input, this.identity.public.id),
      q = checked.request;
    return this.run((values, record) => {
      const found = registry.lookup(
        record,
        this.identity.public.id,
        q.sequence,
        q.operationId,
      );
      if (found.operation) {
        if (found.operation.fingerprint !== checked.fingerprint)
          throw Error("Pedido de recurso repetido com outros valores");
        return found.operation;
      }
      if (
        found.retired ||
        q.sequence !== record.nextSequence ||
        record.operations.some((op) => op.phase === "copy-pending")
      )
        throw Error("Recurso pendente, pedido retirado ou sequência diferente");
      const readers = resolveReaders(q.recipients);
      if (
        readers !== "public" &&
        (readers.length > 64 ||
          readers.some((c) => !validateIdentity(c)) ||
          new Set(readers.map((c) => c.id)).size !== readers.length)
      )
        throw Error("Leitores do recurso inválidos");
      const scope =
        readers === "public" ? "public" : readers.map((c) => c.id).sort();
      if (!same(scope, q.recipients))
        throw Error("Os cartões não correspondem aos leitores do recurso");
      const bundle = createBundle(
        this.identity,
        "site-resource",
        q.content,
        readers,
        q.ttlMs,
      );
      verifyBundle(bundle);
      const ref = describeSiteResource(q.content, {
        id: bundle.manifest.id,
        authorId: bundle.manifest.author.id,
        kind: bundle.manifest.kind,
      });
      const prepared = registry.prepare(record, this.identity.public.id, q, {
        reference: ref,
        created: bundle.manifest.created,
        expires: bundle.manifest.expires,
        bundleHash: hash(canonical(bundle)),
        recipients: scope,
      });
      this.verify(bundle, prepared.operation);
      values.write(this.key + ":stage", bundle);
      values.write(this.key + ":record", prepared.record);
      return prepared.operation;
    });
  }
  authorizedBundle(handle: ResourceOperationHandle): Bundle | null {
    return this.run((_values, record, stage) => {
      const { operation } = registry.lookup(
        record,
        this.identity.public.id,
        handle.sequence,
        handle.operationId,
      );
      if (!operation || operation.fingerprint !== handle.fingerprint)
        throw Error("Resultado de criação desconhecido");
      if (operation.phase !== "copy-pending") return null;
      insist(stage !== null, "Preparação de recurso em falta");
      verifyBundle(stage);
      return stage;
    });
  }
  ready(handle: ResourceOperationHandle, copied: Bundle) {
    return this.run((values, record, stage) => {
      const { operation } = registry.lookup(
        record,
        this.identity.public.id,
        handle.sequence,
        handle.operationId,
      );
      if (!operation || operation.fingerprint !== handle.fingerprint)
        throw Error("Resultado de criação desconhecido");
      this.verify(copied, operation);
      if (operation.phase === "ready") return operation;
      insist(
        operation.phase === "copy-pending" &&
          stage !== null &&
          same(stage, copied),
        "A cópia não corresponde ao recurso autorizado",
      );
      verifyBundle(copied);
      const next = registry.ready(
        record,
        this.identity.public.id,
        handle.sequence,
        handle.operationId,
        handle.fingerprint,
        hash(canonical(copied)),
      );
      values.write(this.key + ":record", next);
      values.remove(this.key + ":stage");
      return next.operations.find((op) => op.sequence === handle.sequence)!;
    });
  }
}

File: native/sites/resource_catalog.go
package sites

import (
	"errors"
	"github.com/JohnnyPBelo/relayloom/native/core"
	"github.com/JohnnyPBelo/relayloom/native/groupstore"
	"sort"
	"time"
)

// ResourceCatalog journals signing-owned creation, never automatic broadcast.
// The runtime remains responsible for copying and verifying the exact bundle.
type ResourceCatalog struct {
	database CatalogDatabase
	identity core.Identity
	key      string
	now      func() int64
}
type ResourceHandle struct {
	Sequence    int64  `json:"sequence"`
	OperationID string `json:"operationId"`
	Fingerprint string `json:"fingerprint"`
}

func NewResourceCatalog(database CatalogDatabase, identity core.Identity) *ResourceCatalog {
	key, _ := core.Canonical(map[string]any{"domain": "relayloom/resource-catalog-key/1", "owner": identity.Public.ID})
	return &ResourceCatalog{database, identity, "resource:" + core.Hash(key), func() int64 { return time.Now().UnixMilli() }}
}
func resourceHandle(op map[string]any) ResourceHandle {
	sequence, _ := creationNumber(op["sequence"])
	return ResourceHandle{sequence, docTextValue(op["operationId"]), docTextValue(op["fingerprint"])}
}
func (c *ResourceCatalog) verify(bundle core.Bundle, op map[string]any) error {
	if err := core.VerifyBundleAt(bundle, bundle.Manifest.Created); err != nil {
		return err
	}
	if bundle.Manifest.Kind != "site-resource" || bundle.Manifest.Author.ID != c.identity.Public.ID {
		return creationError()
	}
	value, err := core.DecryptBundleAt(bundle, &c.identity, bundle.Manifest.Created)
	if err != nil {
		return err
	}
	resource, err := ParseResource(value)
	if err != nil {
		return err
	}
	ref, err := DescribeResource(resource, map[string]any{"id": bundle.Manifest.ID, "authorId": bundle.Manifest.Author.ID, "kind": bundle.Manifest.Kind})
	if err != nil {
		return err
	}
	created, _ := creationNumber(op["created"])
	expires, _ := creationNumber(op["expires"])
	encoded, err := core.Canonical(bundle)
	if err != nil {
		return err
	}
	if !creationEqual(ref, op["reference"]) || !creationEqual(readersOf(bundle), op["recipients"]) || bundle.Manifest.Created != created || bundle.Manifest.Expires != expires || core.Hash(encoded) != op["bundleHash"] {
		return creationError()
	}
	return nil
}
func (c *ResourceCatalog) run(callback func(*PrivateRecords, map[string]any, *core.Bundle) error) error {
	return c.database.Update(func(tx *groupstore.Tx) error {
		return RunResourcePrivate(tx, c.identity, func(values *PrivateRecords) error {
			saved, found, err := values.Read(c.key + ":record")
			if err != nil {
				return err
			}
			raw, staged, err := values.Read(c.key + ":stage")
			if err != nil {
				return err
			}
			var record map[string]any
			if !found {
				keys, err := tx.Keys("resource:")
				if err != nil {
					return err
				}
				if len(keys) != 0 {
					return privateIntegrity("registo de recursos ausente com dados existentes")
				}
				record, err = InitialResourceRecord(c.identity.Public.ID)
				if err != nil {
					return err
				}
			} else {
				record, err = ValidateResourceRecord(saved, c.identity.Public.ID)
				if err != nil {
					return privateIntegrity("registo de criação de recursos inválido")
				}
			}
			var pending map[string]any
			for _, item := range record["operations"].([]any) {
				op := item.(map[string]any)
				if op["phase"] == "copy-pending" {
					pending = op
				}
			}
			var bundle *core.Bundle
			if pending != nil {
				if !staged {
					return privateIntegrity("preparação de recurso em falta")
				}
				encoded, err := core.Canonical(raw)
				if err != nil {
					return privateIntegrity("preparação ilegível")
				}
				decoded, err := core.DecodeBundle(encoded)
				if err != nil || c.verify(decoded, pending) != nil {
					return privateIntegrity("preparação de recurso inválida")
				}
				bundle = &decoded
				if decoded.Manifest.Expires <= c.now() {
					record, err = ExpireResourceCreation(record, c.identity.Public.ID, c.now())
					if err != nil {
						return err
					}
					if err = values.Remove(c.key + ":stage"); err != nil {
						return err
					}
					if err = values.Write(c.key+":record", record); err != nil {
						return err
					}
					bundle = nil
				}
			} else if staged {
				return privateIntegrity("assinatura sem operação pendente")
			}
			return callback(values, record, bundle)
		})
	})
}
func (c *ResourceCatalog) State() (map[string]any, error) {
	var result map[string]any
	err := c.run(func(_ *PrivateRecords, r map[string]any, _ *core.Bundle) error {
		var err error
		result, err = creationClone(r, ResourceRecordBytes)
		return err
	})
	return result, err
}
func (c *ResourceCatalog) Operation(sequence int64, id string) (map[string]any, bool, error) {
	var result map[string]any
	var retired bool
	err := c.run(func(_ *PrivateRecords, r map[string]any, _ *core.Bundle) error {
		var err error
		_, result, retired, err = LookupResourceOperation(r, c.identity.Public.ID, sequence, id)
		return err
	})
	return result, retired, err
}

// resolve is called only for a new operation; retries don't depend on changed
// contact cards and never create another signature with a fresh nonce.
func (c *ResourceCatalog) Prepare(input any, resolve func(any) ([]core.PublicIdentity, bool, error)) (map[string]any, error) {
	q, fingerprint, err := ResourceCreationRequest(input, c.identity.Public.ID)
	if err != nil {
		return nil, err
	}
	sequence, _ := creationNumber(q["sequence"])
	id := docTextValue(q["operationId"])
	var result map[string]any
	err = c.run(func(values *PrivateRecords, record map[string]any, _ *core.Bundle) error {
		_, prior, retired, err := LookupResourceOperation(record, c.identity.Public.ID, sequence, id)
		if err != nil {
			return err
		}
		if prior != nil {
			if prior["fingerprint"] != fingerprint {
				return errors.New("pedido de recurso repetido com outros valores")
			}
			result = prior
			return nil
		}
		next, _ := creationNumber(record["nextSequence"])
		if retired || sequence != next {
			return creationError()
		}
		for _, raw := range record["operations"].([]any) {
			if raw.(map[string]any)["phase"] == "copy-pending" {
				return errors.New("recurso pendente")
			}
		}
		cards, public, err := resolve(q["recipients"])
		if err != nil {
			return err
		}
		var scope any = "public"
		if !public {
			if len(cards) < 1 || len(cards) > 64 {
				return creationError()
			}
			ids := []string{}
			seen := map[string]bool{}
			for _, card := range cards {
				if core.ValidateIdentity(card) != nil || seen[card.ID] {
					return creationError()
				}
				ids = append(ids, card.ID)
				seen[card.ID] = true
			}
			sort.Strings(ids)
			scope = ids
		}
		if !creationEqual(scope, q["recipients"]) {
			return errors.New("cartões diferentes dos leitores")
		}
		ttl, _ := creationNumber(q["ttlMs"])
		bundle, err := core.CreateBundleAt(c.identity, "site-resource", q["content"], cards, public, ttl, c.now())
		if err != nil {
			return err
		}
		ref, err := DescribeResource(q["content"], map[string]any{"id": bundle.Manifest.ID, "authorId": bundle.Manifest.Author.ID, "kind": bundle.Manifest.Kind})
		if err != nil {
			return err
		}
		encoded, err := core.Canonical(bundle)
		if err != nil {
			return err
		}
		updated, op, err := PrepareResourceCreation(record, c.identity.Public.ID, q, map[string]any{"reference": ref, "created": bundle.Manifest.Created, "expires": bundle.Manifest.Expires, "bundleHash": core.Hash(encoded), "recipients": q["recipients"]})
		if err != nil {
			return err
		}
		if err = c.verify(bundle, op); err != nil {
			return err
		}
		if err = values.Write(c.key+":stage", bundle); err != nil {
			return err
		}
		if err = values.Write(c.key+":record", updated); err != nil {
			return err
		}
		result = op
		return nil
	})
	return result, err
}
func (c *ResourceCatalog) AuthorizedBundle(handle ResourceHandle) (*core.Bundle, error) {
	var result *core.Bundle
	err := c.run(func(_ *PrivateRecords, record map[string]any, stage *core.Bundle) error {
		_, op, _, err := LookupResourceOperation(record, c.identity.Public.ID, handle.Sequence, handle.OperationID)
		if err != nil {
			return err
		}
		if op == nil || op["fingerprint"] != handle.Fingerprint {
			return errors.New("resultado de criação desconhecido")
		}
		if op["phase"] != "copy-pending" {
			return nil
		}
		if stage == nil {
			return privateIntegrity("preparação em falta")
		}
		if err = core.VerifyBundleAt(*stage, c.now()); err != nil {
			return err
		}
		result = stage
		return nil
	})
	return result, err
}
func (c *ResourceCatalog) Ready(handle ResourceHandle, copied core.Bundle) (map[string]any, error) {
	var result map[string]any
	err := c.run(func(values *PrivateRecords, record map[string]any, stage *core.Bundle) error {
		_, op, _, err := LookupResourceOperation(record, c.identity.Public.ID, handle.Sequence, handle.OperationID)
		if err != nil {
			return err
		}
		if op == nil || op["fingerprint"] != handle.Fingerprint {
			return errors.New("resultado de criação desconhecido")
		}
		if err = c.verify(copied, op); err != nil {
			return err
		}
		if op["phase"] == "ready" {
			result = op
			return nil
		}
		if op["phase"] != "copy-pending" || stage == nil || !creationEqual(*stage, copied) {
			return errors.New("cópia diferente da preparação")
		}
		if err = core.VerifyBundleAt(copied, c.now()); err != nil {
			return err
		}
		data, err := core.Canonical(copied)
		if err != nil {
			return err
		}
		updated, err := ResourceCopyReady(record, c.identity.Public.ID, handle.Sequence, handle.OperationID, handle.Fingerprint, core.Hash(data))
		if err != nil {
			return err
		}
		if err = values.Write(c.key+":record", updated); err != nil {
			return err
		}
		if err = values.Remove(c.key + ":stage"); err != nil {
			return err
		}
		_, result, _, err = LookupResourceOperation(updated, c.identity.Public.ID, handle.Sequence, handle.OperationID)
		return err
	})
	return result, err
}
